File: src/lib/fidl/rust_next/fidl_next_protocol/src/service.rs
// Copyright 2024 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use core::error::Error;

use crate::Transport;

/// An instance of a FIDL service.
pub trait ServiceInstance {
    /// The error type for the instance.
    type Error: Error + Send + Sync + 'static;

    /// The transport type created by connecting to a member.
    type Transport: Transport;

    /// Attempts to connect the given member.
    fn connect(&mut self, member: &str) -> Result<Self::Transport, Self::Error>;
}

File: src/storage/lib/vfs/cpp/service.cc
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/storage/lib/vfs/cpp/service.h"

#include <fidl/fuchsia.io/cpp/common_types.h>
#include <lib/zx/channel.h>
#include <zircon/errors.h>
#include <zircon/types.h>

#include <utility>

namespace fs {

Service::Service(Connector connector) : connector_(std::move(connector)) {}

Service::~Service() = default;

fuchsia_io::NodeProtocolKinds Service::GetProtocols() const {
  return fuchsia_io::NodeProtocolKinds::kConnector;
}

zx_status_t Service::ConnectService(zx::channel channel) {
  if (!connector_) {
    return ZX_ERR_NOT_SUPPORTED;
  }
  return connector_(std::move(channel));
}

}  // namespace fs
